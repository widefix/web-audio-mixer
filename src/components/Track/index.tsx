import React, { useEffect, useRef, useState } from 'react'
import classnames from 'classnames'
import { keys } from 'ramda'

import Fader from './../Fader'
import Knob from '../Knob'

import style from './style.module.css'
import { getAverage } from '../MasterTrack/helpers'
import { useRecoilValue } from 'recoil'
import { statusState } from '../../store/playback'

const DEFAULT_VOLUME = 70

interface TrackProps {
  analyser: AnalyserNode
  id: string
  title: string
  volume: number
  pan: number
  isMuted: boolean
  isSolo: boolean
  isEffectsDisabled: boolean
  // TODO: Types
  fx: any
  onMute: (trackId) => {}
  onSolo: (trackId) => {}
  onBypass: (trackId) => {}
  onVolumeChange: (trackId, volume) => any
  onSendLevelChange?: (trackId, id, value) => void
  onPanChange: (trackId, value) => {}
  hasSolo: boolean
}

const Track: React.FC<TrackProps> = (props) => {
  const [soloMutePriority, setSoloMutePriority] = useState('')

  const setSoloMutePriorityHelper = (value: 's' | 'm') => {
    if (soloMutePriority.includes(value)) {
      setSoloMutePriority(soloMutePriority.replace(value, ''))
    } else if (soloMutePriority.length < 2) {
      setSoloMutePriority(soloMutePriority + value)
    } else {
      setSoloMutePriority(value)
    }
  }

  const status = useRecoilValue(statusState)

  const analyserDiv = useRef<HTMLDivElement>()

  const height = 176
  const array = new Uint8Array(props.analyser.frequencyBinCount)
  props.analyser.getByteFrequencyData(array)

  const average = useRef(0)

  let frame
  const drawMeter = () => {
    props.analyser.getByteFrequencyData(array)
    average.current = getAverage(array)
    if (analyserDiv.current && analyserDiv.current.style) {
      const maxValueToCurrentVolume = height * (props.volume / 100)
      const progress = (height / 100) * average.current
      analyserDiv.current.style.height = `${Math.min(progress, maxValueToCurrentVolume)}px`
    }

    frame = requestAnimationFrame(drawMeter)
  }

  const toZero = () => {
    if (analyserDiv.current && analyserDiv.current.style) {
      if (average.current > 0) {
        average.current -= 2
        analyserDiv.current.style.height = `${(height / 100) * average.current}px`
        frame = requestAnimationFrame(toZero)
      } else {
        analyserDiv.current.style.height = `${(height / 100) * 0}px`
        cancelAnimationFrame(frame)
      }
    }
  }

  useEffect(() => {
    if  (status === 'PLAYING') {
      frame = requestAnimationFrame(drawMeter)
    } else {
      toZero()
    }
    return () => cancelAnimationFrame(frame)
  }, [status, props.volume])

  return (
    <div className={style.track}>

      {(!props.hasSolo || props.isSolo) && <div className={style.meterValue} ref={analyserDiv} />}

      {keys(props.fx).length > 0 && (
        <button
          className={classnames(
            style.button,
            props.isEffectsDisabled && style.isPressed
          )}
          onClick={() => props.onBypass(props.id)}
        >
          Bypass FX
        </button>
      )}
      <div className={style.trackControls}>
        <button
          className={classnames(
            style.button,
            style.soloButton,
            props.isSolo && style.isPressed
          )}
          onClick={() => {
            setSoloMutePriorityHelper('s')
            if (soloMutePriority === 'm') {
              props.onMute(props.id)
              props.onSolo(props.id)
            } else if (soloMutePriority === 'ms') {
              props.onSolo(props.id)
              props.onMute(props.id)
            } else if (soloMutePriority === 'sm') {
              props.onSolo(props.id)
            } else {
              props.onSolo(props.id)
            }
          }}
        >
          S
        </button>
        <div
          className={classnames(
            style.buttonSeparator,
            props.isMuted || props.isSolo ? style.isHidden : ''
          )}
        />
        {!props.hasSolo && (
          <button
            className={classnames(style.button, props.isMuted && style.isPressed)}
            onClick={() => {
              setSoloMutePriorityHelper('m')
              props.onMute(props.id)
            }}
          >
            M
          </button>
        )}
      </div>

      <Knob
        min={0}
        max={100}
        value={props.pan}
        onChange={(value) => props.onPanChange(props.id, value)}
      />

      {keys(props.fx).length > 0 && (
        <div className={style.sends}>
          {keys(props.fx).map((sendId) => (
            <div className={style.send} key={sendId}>
              <span className={style.sendTitle}>{sendId}</span>
              <Fader
                onChange={(value) => props.onSendLevelChange(props.id, sendId, value)}
                value={props.fx[sendId]}
              />
            </div>
          ))}
        </div>
      )}

      <Fader
        onChange={(value) => props.onVolumeChange(props.id, value)}
        value={props.volume}
        onDoubleClick={() => props.onVolumeChange(props.id, DEFAULT_VOLUME)}
        isVertical
      />

      <div className={style.title}>
        <span className={style.titleText}>{props.title}</span>
      </div>
    </div>
  )
}

Track.defaultProps = {
  // TODO: Default track props do not work
  title: 'Untitled',
  volume: DEFAULT_VOLUME,
  isMuted: false,
  isSolo: false,
  isEffectsDisabled: false,
  // TODO: why is it always truthy even though no send is passed
  fx: null,
  onMute: () => null,
  onSolo: () => null,
  onBypass: () => null,
  onVolumeChange: () => null,
  onSendLevelChange: () => null,
  hasSolo: false
} as Partial<TrackProps>

export default Track
