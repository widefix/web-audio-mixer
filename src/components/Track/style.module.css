.track {
  display: flex;
  flex-flow: column;
  width: 80px;
  min-width: 80px;
  max-width: 80px;
  min-height: 380px;
  text-align: center;
  position: relative;

  background: #808080;
  background-size: contain;
}

.trackControls {
  position: relative;
  display: flex;
  flex-direction: row-reverse;
  margin-bottom: 16px;
  background-color: #454545;
}

.trackControls button {
  flex: 0 0 50%;
}

.track:first-child {
  margin-left: auto;
}

.track:not(:last-child) {
  margin-right: 8px;
}

.button {
  font-size: 1em;
  font-weight: 700;
  flex: 1 1 auto;

  color: #fff;
  overflow: hidden;

  padding: 7px 8px;

  cursor: pointer;
  transition: all 80ms ease-in-out;
  vertical-align: middle;
  white-space: nowrap;
  text-overflow: ellipsis;

  border: none;
  outline: none;
  background-color: #454545;
  position: relative;
}

.buttonSeparator {
  width: 1px;
  height: 20px;
  background-color: #ffffff;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translateY(-50%);
  visibility: visible;
}

.isHidden {
  visibility: hidden;
}

.button:hover {
  transform: translateY(1px);
}

.isPressed {
  background-color: #e728ea;
}

.soloButton.isPressed {
  background-color: #49c7f2;
}

.mute,
.reset,
.bypass {
  position: relative;

  display: flex;

  margin: 8px 8px 16px;
}

.sends {
  font-size: 0.8rem;

  margin-top: 10px;
}

.send {
  width: 60%;
  height: 26px;
  margin: 0 auto 10px;
}

.sendTitle {
  font-weight: bold;

  display: inline-block;
  overflow: hidden;

  width: 100%;

  text-align: left;
  white-space: nowrap;
  text-overflow: ellipsis;

  color: #555;
  text-shadow: 1px 1px 0 #eee;
}

.title {
  line-height: 44px;
  min-height: 44px;
  color: #fff;
  font-weight: 700;
  font-size: 10px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 8px;
}

.titleText {
  white-space: nowrap;
}

.meterValue {
    position: absolute;
    left: 50%;
    margin-left: -3px;
    bottom: 70px;
    width: 6px;
    z-index: 1;
    overflow: hidden;
    will-change: height;
}

.meterValue::after {
  content: '';
  height: 176px;
  bottom: 0;
  width: 6px;
  position: absolute;
  border-radius: 8px;
  left: 0;
  background: linear-gradient(0deg, rgba(40,234,182,1) 0%, rgba(255,231,0,1) 60%, rgba(255,0,0,1) 100%);
}
