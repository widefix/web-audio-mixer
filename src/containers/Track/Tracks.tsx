import React, {useEffect} from 'react'

import Track from '../../components/Track'
import { useRecoilState, useRecoilValue } from 'recoil'
import { isSoloSelector, tracksState } from '../../store/store'
import style from '../../components/Desk/style.module.css'
import { useMxContext } from "../../hooks"
import { UserInputEvents } from "../Mixdesk"

const Tracks = () => {
  const [tracks, setTracks ]= useRecoilState(tracksState)
  const hasSolo = useRecoilValue(isSoloSelector)

  const {mx, isReady, mixerChanges, onUserInput} = useMxContext()

  const toggleTrackSolo = async (trackId) => {
    if (mx.current) {
      await mx.current.soloTrack(trackId)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        onUserInput({type: UserInputEvents.playingStateChange, value: !track.isSolo ? "solo" : null, trackId: track.id})
        return {
          ...track,
          isSolo: !track.isSolo,
        };
      }

      return track
    }))
  }

  const toggleTrack = async (trackId) => {
    if (mx.current) {
      await mx.current.toggleTrack(trackId)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        onUserInput({type: UserInputEvents.playingStateChange, value: !track.isMuted ? "mute" : null, trackId: track.id})
        return {
          ...track,
          isMuted: !track.isMuted,
        };
      }

      return track
    }))
  }

  const toggleTrackFx = async (trackId) => {
    if (mx.current) {
      await mx.current.toggleTrackFx(trackId)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        return {
          ...track,
          isEffectsDisabled: !track.isEffectsDisabled,
        };
      }

      return track;
    }))
  }

  const setTrackVolume = async (trackId, value) => {
    if (mx.current) {
      await mx.current.setTrackVolume(trackId, value)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        onUserInput({type: UserInputEvents.trackVolumeChange, value: value, trackId: track.id})
        return {
          ...track,
          volume: value,
        };
      }
      return track;
    }))
  }

  const setTrackSendLevel = async (trackId, fxId, value) => {
    if (mx.current) {
      await mx.current.setTrackSendLevel(trackId, fxId, value)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        return {
          ...track,
          send: {
            ...track.send,
            [fxId]: value,
          },
        };
      }

      return track;
    }))
  }

  const setTrackPan = async (trackId, value) => {
    if (mx.current) {
      await mx.current.setTrackPan(trackId, value)
    }

    setTracks((tracks) => tracks.map(track => {
      if (track.id === trackId) {
        onUserInput({type: UserInputEvents.panChange, value: value, trackId: track.id})
        return {
          ...track,
          pan: value,
        };
      }
      return track;
    }))
  }

  // Loading mixer state between song key changes
  useEffect(() => {
    if (mixerChanges.tracks && isReady) {
      for (const [trackId, track] of Object.entries(mixerChanges.tracks)) {
        if (track?.volume) {
          setTrackVolume(trackId, track.volume)
        }
        if (track?.pan) {
          setTrackPan(trackId, track.pan)
        }
        switch (track?.playingState) {
          case "mute":
            toggleTrack(trackId)
            break
          case "solo":
            toggleTrackSolo(trackId)
            break
          default:
            break
        }
      }
    }
  }, [mixerChanges.songKey, isReady])

  return (
    <div className={style.tracks}>
      {isReady && tracks.map((track, index) => (
        <Track
          key={track.id + index}
          {...track}
          hasSolo={hasSolo}
          onSolo={toggleTrackSolo}
          onMute={toggleTrack}
          onBypass={toggleTrackFx}
          onVolumeChange={setTrackVolume}
          onSendLevelChange={setTrackSendLevel}
          onPanChange={setTrackPan}
        />
      ))}
    </div>
  )
}

export default Tracks
