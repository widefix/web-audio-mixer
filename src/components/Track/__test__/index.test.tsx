import React from 'react'
import { shallow, mount } from 'enzyme'

import Track from '..'

const propsMock = {
  id: 'track',
  title: 'Track',
  volume: 42.0,
  isSolo: false,
  isMuted: false,
  isEffectsDisabled: false,
  pan: 50,
  fx: {
    delay: 0,
    reverb: 0
  },
  onMute: jest.fn(),
  onBypass: jest.fn(),
  onVolumeChange: jest.fn(),
  onSendLevelChange: jest.fn(),
  onSolo: jest.fn(),
  onPanChange: jest.fn()
}

const MUTE_BUTTON = 1
const SOLO_BUTTON = 2
const FX_BYPASS_BUTTON = 0
const FADER_VOLUME = 2
const FADER_SEND = 1

describe('<Track />', () => {
  it('should renders without any errors', () => {
    const wrapper = shallow(<Track {...propsMock} />)

    expect(wrapper).toMatchSnapshot()
  })

  describe('onMute', () => {
    it('should be called on mute toggler click', () => {
      const wrapper = shallow(<Track {...propsMock} />)
      wrapper.find('button').at(MUTE_BUTTON).simulate('click')

      expect(propsMock.onMute).toBeCalled()
    })
  })

  describe('onSolo', () => {
    it('should be called on solo toggler click', () => {
      const wrapper = shallow(<Track {...propsMock} />)

      wrapper.find('button').at(SOLO_BUTTON).simulate('click')

      expect(propsMock.onSolo).toBeCalled()
    })
  })

  describe('onBypass', () => {
    it('should be called on fx bypass toggler click', () => {
      const wrapper = shallow(<Track {...propsMock} />)

      wrapper.find('button').at(FX_BYPASS_BUTTON).simulate('click')

      expect(propsMock.onBypass).toBeCalled()
    })
  })

  describe('onVolumeChange', () => {
    it('should be called on volume change', () => {
      const wrapper = mount(<Track {...propsMock} />)
      wrapper
        .find('Fader')
        .at(FADER_VOLUME)
        .find('FaderThumb')
        .simulate('mouseDown')

      window.document.documentElement.dispatchEvent(new Event('mousemove'))
      window.document.documentElement.dispatchEvent(new Event('mouseup'))

      expect(propsMock.onVolumeChange).toBeCalled()
    })
  })

  describe('onSendLevelChange', () => {
    it('should be called on send level change', () => {
      const wrapper = mount(<Track {...propsMock} />)

      wrapper
        .find('Fader')
        .at(FADER_SEND)
        .find('FaderThumb')
        .simulate('mouseDown')

      window.document.documentElement.dispatchEvent(new Event('mousemove'))
      window.document.documentElement.dispatchEvent(new Event('mouseup'))

      expect(propsMock.onSendLevelChange).toBeCalled()
    })
  })
})
