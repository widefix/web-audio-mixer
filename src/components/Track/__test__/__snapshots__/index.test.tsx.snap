// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> should renders without any errors 1`] = `
<div
  className="track"
>
  <button
    className="button"
    onClick={[Function]}
  >
    Bypass FX
  </button>
  <div
    className="trackControls"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      M
    </button>
    <div
      className="buttonSeparator"
    />
    <button
      className="button"
      onClick={[Function]}
    >
      S
    </button>
  </div>
  <Knob
    degrees={180}
    max={100}
    min={1}
    size={50}
    value={50}
  />
  <div
    className="sends"
  >
    <div
      className="send"
      key="delay"
    >
      <span
        className="sendTitle"
      >
        delay
      </span>
      <Fader
        value={0}
      />
    </div>
    <div
      className="send"
      key="reverb"
    >
      <span
        className="sendTitle"
      >
        reverb
      </span>
      <Fader
        value={0}
      />
    </div>
  </div>
  <Fader
    isVertical={true}
    onChange={[MockFunction]}
    value={42}
  />
  <div
    className="title"
  >
    Track
  </div>
</div>
`;
